src/metrics.rs

```diff
@@ -339,19 +368,61 @@ impl MetricsLayer {
         MetricsLayer {
             meter,
             instruments: Default::default(),
+            callsites: Callsites::new(),
         }
     }
+
+    fn has_metrics_fields(&self, field_set: &FieldSet) -> bool {
+        field_set.iter().any(|field| {
+            let name = field.name();
+            name.starts_with(METRIC_PREFIX_COUNTER)
+                || name.starts_with(METRIC_PREFIX_MONOTONIC_COUNTER)
+                || name.starts_with(METRIC_PREFIX_HISTOGRAM)
+        })
+    }
 }
 
 impl<S> Layer<S> for MetricsLayer
 where
     S: Subscriber + for<'span> LookupSpan<'span>,
 {
     fn on_event(&self, event: &tracing::Event<'_>, _ctx: Context<'_, S>) {
-        let mut metric_visitor = MetricVisitor {
-            instruments: &self.instruments,
-            meter: &self.meter,
-        };
-        event.record(&mut metric_visitor);
+        if self
+            .callsites
+            .has_metrics_field(&event.metadata().callsite())
+        {
+            let mut attributes = Vec::new();
+            let mut visited_metrics = Vec::new();
```

Comment:
```rust
tracing::info!(
  counter.foo = 1,
  counter.bar = 2,
)
```

Since multiple metrics may be output as described above, Vec is used.

I also thought about using thread_local!  to reduce the number of allocations.